lighthouse-core/gather/computed/main-resource.js

```diff
@@ -21,18 +22,18 @@ class MainResource extends ComputedArtifact {
    * @param {LH.ComputedArtifacts} artifacts
    * @return {Promise<LH.WebInspector.NetworkRequest>}
    */
-  compute_(data, artifacts) {
-    const {URL, devtoolsLog} = data;
-    return artifacts.requestNetworkRecords(devtoolsLog)
-      .then(requests => {
-        const mainResource = requests.find(request => request.url === URL.finalUrl);
+  async compute_(data, artifacts) {
+    const finalUrl = data.URL.finalUrl;
```

Comment:
Caught exception: Cannot read property 'finalUrl' of undefined

Reply:
@janerikmai i can't reproduce this. what audit is calling into requestMainResource when you get this result?

Reply:
Yeah this PR doesn't touch the structure of data, it's been unchanged and we have 0 instances of `Cannot read property 'finalUrl' of undefined` in Sentry 🤔

Reply:
Exception in `critical-request-chains`, `redirects`, `uses-rel-preload`, `http-status-code`,  `is-crawlable`, `hreflang` and `canonical` audits. It is working with my local fix 
`const mainResource = requests.find(request => request.url === URL.finalUrl.split('#')[0]);` so i guess this Exception was introduced by changing `compute` to `async compute` and await, so my guess is that compute with async is faster than the finalUrl node is appended to data.

Reply:
This works without exception =>
```javascript
/**
 * @license Copyright 2017 Google Inc. All Rights Reserved.
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
 */
'use strict';

const ComputedArtifact = require('./computed-artifact');
const URL = require('../../lib/url-shim');


/**
 * @fileoverview This artifact identifies the main resource on the page. Current solution assumes
 * that the main resource is the first non-rediected one.
 */
class MainResource extends ComputedArtifact {
  get name() {
    return 'MainResource';
  }

  /**
   * @param {!DevtoolsLog} devtoolsLog
   * @param {!ComputedArtifacts} artifacts
   * @return {!WebInspector.NetworkRequest}
   */
  compute_(devtoolsLog, artifacts) {
    return artifacts.requestNetworkRecords(devtoolsLog)
      .then(requests => {
        const mainResource = requests.find(request => artifacts.URL.finalUrl.startsWith(request.url) && URL.equalWithExcludedFragments(request.url, artifacts.URL.finalUrl));

        if (!mainResource) {
          throw new Error('Unable to identify the main resource');
        }

        return mainResource;
      });
  }
}

module.exports = MainResource;
```

Reply:
Ah, there have been many other changes to this since 2.9.4. This is a patch on 3.x. `requestMainResource` doesn't take `devtoolsLog` anymore, it takes `{URL, devtoolsLog}`

Reply:
ok i will upgrade thx :)